qabelbox/src/main/java/de/qabel/qabelbox/box/provider/BoxProvider.kt

```diff
@@ -0,0 +1,221 @@
+package de.qabel.qabelbox.box.provider
+
+import android.content.BroadcastReceiver
+import android.content.Context
+import android.content.Intent
+import android.content.IntentFilter
+import android.database.Cursor
+import android.database.MatrixCursor
+import android.os.CancellationSignal
+import android.os.Handler
+import android.os.ParcelFileDescriptor
+import android.provider.DocumentsContract
+import android.provider.DocumentsContract.Document
+import android.provider.DocumentsContract.Root
+import android.provider.DocumentsProvider
+import android.provider.MediaStore.Video.Media
+import android.util.Log
+import de.qabel.box.storage.exceptions.QblStorageException
+import de.qabel.qabelbox.BuildConfig
+import de.qabel.qabelbox.QabelBoxApplication
+import de.qabel.qabelbox.QblBroadcastConstants
+import de.qabel.qabelbox.R
+import de.qabel.qabelbox.dagger.modules.BoxModule
+import de.qabel.qabelbox.box.dto.BrowserEntry
+import de.qabel.qabelbox.box.dto.ProviderUpload
+import de.qabel.qabelbox.box.dto.UploadSource
+import de.qabel.qabelbox.box.interactor.ProviderUseCase
+import de.qabel.qabelbox.dagger.components.DaggerBoxComponent
+import rx.lang.kotlin.firstOrNull
+import java.io.File
+import java.io.FileNotFoundException
+import java.net.URLConnection
+import java.util.*
+import javax.inject.Inject
+
+open class BoxProvider : DocumentsProvider() {
+
+    @Inject
+    lateinit var useCase: ProviderUseCase
+    open val handler by lazy { Handler(context.mainLooper) }
+
+    private val volumesChangedBroadcastReceiver = object : BroadcastReceiver() {
+        override fun onReceive(context: Context, intent: Intent) {
+            notifyRootsUpdated()
+        }
+    }
+
+    override fun onCreate(): Boolean {
+        inject()
+        context.registerReceiver(volumesChangedBroadcastReceiver,
+                IntentFilter(QblBroadcastConstants.Storage.BOX_VOLUMES_CHANGES))
+        return true
+    }
+
+    open fun inject() {
+        val boxComponent = DaggerBoxComponent.builder().boxModule(BoxModule()).build()
+        boxComponent.inject(this)
+    }
+
+    /**
+     * Notify the system that the roots have changed
+     * This happens if identities or prefixes changed.
+     */
+    fun notifyRootsUpdated() {
+        context.contentResolver.notifyChange(DocumentsContract.buildRootsUri(
+                BuildConfig.APPLICATION_ID + AUTHORITY), null)
+    }
+
+    @Throws(FileNotFoundException::class)
+    override fun queryRoots(projection: Array<String>?): Cursor {
+        val netProjection = reduceProjection(projection, DEFAULT_ROOT_PROJECTION)
+        val result = MatrixCursor(netProjection)
+        useCase.availableRoots().forEach {
+            with(it) {
```

Comment:
imho you could leave the `it` in (:= no `with(it)`). It's not long, it would produce fewer lines and less nesting and it would clear which value is from which object